### PR TITLE
allow solos build of individual API driver

### PR DESCRIPTION
**Pull request by khou2020**
add dat2decom utility that supports converting decomposition files into all supported formats
read_decom use driver to read decomposition files, supporting all file formats

### CODE REVIEW

**Review by wkliao — COMMENTED** [review]
This will fail when both PnetCDF and NetCD-4 are enabled.

File: configure.ac [hunk 1]
```diff
@@ -25,7 +25,7 @@ AC_CONFIG_AUX_DIR([./scripts])
 
 m4_ifdef([AC_CONFIG_MACRO_DIRS], [AC_CONFIG_MACRO_DIRS([m4])], [AC_CONFIG_MACRO_DIR([m4])])
 
-AM_INIT_AUTOMAKE([foreign])
+AM_INIT_AUTOMAKE([foreign subdir-objects])
```
Comment (wkliao) [inline configure.ac:28]: Why is this necessary?

**Comment by wkliao** [comment]
Please update utils/README.md for the new program dat2decom.

File: configure.ac (hunk 1 above)
Comment (khou2020) [reply configure.ac:28]: We link e3sm_io_driver.la at utils folder. autoreconf showing a warning message asking it to eb enabled.

Comment (wkliao) [reply configure.ac:28]: PnetCDF does the same. Did you see the same warning? See this line.
https://github.com/Parallel-NetCDF/PnetCDF/blob/8f583ffb01116ba626622809e2267d2ac980fa7c/src/utils/ncmpidiff/Makefile.am#L15

**Comment by khou2020** [comment]
We use nc_type for external types. To enable build without nc, we have to use another class of datatype for the driver.
We can either use MPI type or define our own enum.

**Comment by wkliao** [comment]
> We use nc_type for external types. To enable build without nc, we have to use another class of datatype for the driver.
> We can either use MPI type or define our own enum.

We already have a mapping of those types. See if you can reuse them without much changes to the source codes.

**Comment by khou2020** [comment]
> > We use nc_type for external types. To enable build without nc, we have to use another class of datatype for the driver.
> > We can either use MPI type or define our own enum.
> 
> We already have a mapping of those types. See if you can reuse them without much changes to the source codes.

An easy approach is to define those constants if nc header is not included.

**Comment by wkliao** [comment]
Got an 'make check' error when building with hdf5 only.
```
 src/e3sm_io -k -a hdf5 -f -1 -r 2 -x canonical -y 2 -o ./test_output/hdf5_canonical_f_case_866x72_16p.h5 ./datasets/f_case_866x72_16p.nc
Error in e3sm_io_driver.cpp line 218 function e3sm_io_get_driver: I/O API is not set
Error in read_decomp.cpp line 162 function read_decomp
Error in e3sm_io.c line 402 function main
```

**Comment by khou2020** [comment]
> 
./datasets/f_case_866x72_16p.nc is netcdf file, but netcdf driver is not enabled, so it cannot select the proper driver for that.
To test hdf5 only, we have to use dat2decom to generate hdf5 formatted decomposition file.

Should I make the change?

**Comment by wkliao** [comment]
Got an error trying to run dat2decom.
```
% cat $SCRATCH/PnetCDF/Github/E3SM-IO/datasets/list_f_case.txt 
piodecomp16tasks16io01dims_ioid_514.dat
piodecomp16tasks16io01dims_ioid_516.dat
piodecomp16tasks16io02dims_ioid_548.dat
% ../utils/dat2decom -a hdf5 -i list_f_case.txt -o f_map.h5
Error: decomp_nprocs=47 mismatches among input files 16
```

**Comment by khou2020** [comment]
> Got an error trying to run dat2decom.
> 
> ```
> % cat $SCRATCH/PnetCDF/Github/E3SM-IO/datasets/list_f_case.txt 
> piodecomp16tasks16io01dims_ioid_514.dat
> piodecomp16tasks16io01dims_ioid_516.dat
> piodecomp16tasks16io02dims_ioid_548.dat
> % ../utils/dat2decom -a hdf5 -i list_f_case.txt -o f_map.h5
> Error: decomp_nprocs=47 mismatches among input files 16
> ```

fixed

**Comment by wkliao** [comment]
Another error.
```
% ../utils/dat2decom -r -a adios -i list_f_case.txt -o f_map.bp

ERROR: attribute D1.lengths/max has been defined and its value cannot be changed, in call to DefineAttribute

adios2_define_attribute_array or adios_define_variable_attribute_array
Error in e3sm_io_driver_adios2.cpp:637: NULL
Error in dat2decom.cpp line 437 function add_decomp
Error in dat2decom.cpp line 666 function main
```

**Comment by wkliao** [comment]
Do we have a utility program that can convert a netcdf decomposition file to h5, nc4, and bp?
For example, it reads f_case_866x72_16p.nc and creates f_case_866x72_16p.h5, f_case_866x72_16p.nc4, or f_case_866x72_16p.bp.
These files can be created and used in 'make check'.

**Comment by khou2020** [comment]
> Do we have a utility program that can convert a netcdf decomposition file to h5, nc4, and bp? For example, it reads f_case_866x72_16p.nc and creates f_case_866x72_16p.h5, f_case_866x72_16p.nc4, or f_case_866x72_16p.bp. These files can be created and used in 'make check'.

No. Do we need one?

**Comment by wkliao** [comment]
In order to test the drivers individually, we need to decomposition files in h5, nc4, and bp formats, but we do not have the .dat files for G and I cases. So, if we have that utility program, it can be used to convert the nc files at the "make check" time, without adding those h5/nc4/bp files into the repo.

**Comment by wkliao** [comment]
Got an error when running hdf5 option.
```
% utils/dat2decom -a hdf5 -i list_f_case.txt -o f_case_866x72_16p.h5
% src/e3sm_io f_case_866x72_16p.h5 -k -o can_F_out.h5 -x canonical -a  hdf5

  #009: ../../hdf5-1.13.0/src/H5Gloc.c line 396 in H5G__loc_find_cb(): object '_LOG' doesn't exist
  #009: ../../hdf5-1.13.0/src/H5Gloc.c line 396 in H5G__loc_find_cb(): object 'D1.raw_nreqs' doesn't exist
  #009: ../../hdf5-1.13.0/src/H5Gloc.c line 396 in H5G__loc_find_cb(): object 'D2.raw_nreqs' doesn't exist
  #009: ../../hdf5-1.13.0/src/H5Gloc.c line 396 in H5G__loc_find_cb(): object 'D3.raw_nreqs' doesn't exist
```

**Comment by wkliao** [comment]
When running this command below, I think it should create a BP file, not a folder.
```
../utils/dat2decom -r -a adios -i list_f_case.txt -o f_map.bp
```

**Comment by khou2020** [comment]
> We need to suppress the error message in compatible() of all drivers. Otherwise I got this when running netcdf4 api.
> 
> ```
> ERROR: ADIOS2 only supports bp format version 3 and above, found 0 version 
> 
> adios2_open
> ```

It is adios internal error message. There is no way to turn it off from outside unless we manipulate stdout and stderr.

**Comment by wkliao** [comment]
Is there a way to check BP file's footer, like HDF5 and NetCDF's signature in the header?
So, there is no need to call adios2_open, just POSIX open.

**Comment by khou2020** [comment]
> ERROR: ADIOS2 only supports

fixed

**Comment by wkliao** [comment]
> When running this command below, I think it should create a BP file, not a folder.
> 
> ```
> ../utils/dat2decom -r -a adios -i list_f_case.txt -o f_map.bp
> ```

Please change this and avoid creating file profiling.json.

**Comment by khou2020** [comment]
> > When running this command below, I think it should create a BP file, not a folder.
> > ```
> > ../utils/dat2decom -r -a adios -i list_f_case.txt -o f_map.bp
> > ```
> 
> Please change this and avoid creating file profiling.json.

That's only for adios1. All adios2 files are folders even if sub-filing is not enabled. There is no way to change it from application.
We can only post process it into a single file.

**Comment by wkliao** [comment]
I can generate various files for F case.
```
% ls -lGgh f_case_866x72_16p*
-rw------- 1 566K Mar 23 10:59 f_case_866x72_16p.bp
-rw------- 1 241K Mar 22 17:27 f_case_866x72_16p.h5
-rw------- 1 235K May 15  2021 f_case_866x72_16p.nc
-rw------- 1 968K Mar 22 17:32 f_case_866x72_16p.nc4
```

Can you generate the files for G and I cases?
I do not have their .dat files.

**Comment by wkliao** [comment]
FYI. I can create nc4 files by using nccopy.
So, we still need h5 and bp files.
